### PR TITLE
Refactor and clean `IR` module

### PR DESCRIPTION
**Pull request by mofeing**
I've refactored the `IR` module so it's easier to maintain and read. I have renamed some functions for more Julia-like names. Also, I have added some missing methods in `IR` which use functionality from the C-API.

One of the biggest changes is that I've renamed `MLIRType` to `Type`. I found out that if we use `Core.Type` inside the `IR` module instead of the implicit `Type`, then there is no symbol clash. In order to avoid other clashes, this type is not exported so we need to access it like `MLIR.IR.Type`.

Some of these changes are opinionated, some of them not. Please, tell me your opinion.

# New features
- Added `Identifier`, `SymbolTable`, `IntegerSet`, `AffineExpr`, `AffineMap` types
- `MLIRType` renamed to `Type` with no clash with `Core.Type`

# To do
- [x] Refactor `Type`
- [x] Refactor `Attribute`
- [x] Refactor examples
- [x] Add `@affinemap` #35

# To do list for other PRs
- Implement `simplify(::AffineMap)`
- Add `MemRefDescriptor`
- Refactor pass infrastructure
- Missing `Attribute`s:
  - `StridedLayout`
  - `ResourceElementsAttr`
- Transform dialect support

### CODE REVIEW

**Comment by jumerckx** [comment]
This is super cool, thank you for this. I'll look into this and try it out the coming days but on cursory glance I don't see anything I would personally object to.
I believe you forgot to check in `Type.jl` and `Attribute.jl` or are you still working on it?

**Comment by mofeing** [comment]
> I believe you forgot to check in `Type.jl` and `Attribute.jl` or are you still working on it?

Yes, I'm still working on them.

**Review by Pangoraw — COMMENTED** [review]
Nice! I left a few comments from skimming through the PR.

As a side note, I think we should not rely so much on asserts to throw errors since I discovered (since writing the initial MLIR.IR) that they can be removed depending on the optimization level so I think we should always perform the checks instead.

File: src/IR/AffineMap.jl [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+Returns the result at the given position.
+"""
+result(map::AffineMap, pos) = AffineExpr(API.mlirAffineMapGetResult(map, pos))
+
+"""
+    ninputs(affineMap)
+
+Returns the number of inputs (dimensions + symbols) of the given affine map.
+"""
+ninputs(map::AffineMap) = API.mlirAffineMapGetNumInputs(map)
```
[the remark this answers is not in the document]
Comment (mofeing) [reply src/IR/AffineMap.jl:167]: yep, I forgot about `num_args`. should be fixed now. thanks for noticing!

File: src/IR/AffineExpr.jl
[the remark this answers is not in the document]
Comment (mofeing) [reply src/IR/AffineExpr.jl:1]: Refactored!

```julia
julia> @macroexpand @affinemap (a,b)[s] -> (a+b+1, s%2)
quote
    #= /Users/mofeing/Developer/MLIR.jl/src/IR/AffineMap.jl:305 =#
    var"#813#a" = (MLIR.IR.IR).AffineDimensionExpr(1)
    var"#814#b" = (MLIR.IR.IR).AffineDimensionExpr(2)
    #= /Users/mofeing/Developer/MLIR.jl/src/IR/AffineMap.jl:306 =#
    var"#815#s" = (MLIR.IR.IR).SymbolExpr(1)
    #= /Users/mofeing/Developer/MLIR.jl/src/IR/AffineMap.jl:308 =#
    MLIR.IR.AffineMap(2, 1, [var"#813#a" + var"#814#b" + (MLIR.IR.ConstantExpr)(1), var"#815#s" % (MLIR.IR.ConstantExpr)(2)])
end
```

File: src/IR/IR.jl [hunk 2]
```diff
@@ -54,14 +54,6 @@ function Base.String(str::API.MlirStringRef)
     Base.unsafe_string(pointer(str.data), str.length)
 end
 
-### Logical Result
-
-mlirLogicalResultSuccess() = API.MlirLogicalResult(1)
-mlirLogicalResultFailure() = API.MlirLogicalResult(0)
-
-mlirLogicalResultIsSuccess(result) = result.value != 0
-mlirLogicalResultIsFailure(result) = result.value == 0
```
[the remark this answers is not in the document]
Comment (mofeing) [reply src/IR/IR.jl]: Ahh it's the same as with `mlirContextIsNull`, `mlirOperationIsNull`, ... Maybe we can write a PR for upstream LLVM to move out those implementations from headers to have symbols exported.

**Comment by mofeing** [comment]
@jumerckx `Type` and `Attribute` now submitted!

> As a side note, I think we should not rely so much on asserts to throw errors since I discovered (since writing the initial MLIR.IR) that they can be removed depending on the optimization level so I think we should always perform the checks instead.

@Pangoraw ahh I thought you wanted asserts specifically because of that: because they can be deactivated. I can replace them for exceptions. Should we throw `ArgumentError`?

I just discovered that the C-API also changes in different versions, so I need to revise and add some checks for each LLVM version. Since we are all using very-up-to-date LLVM builds, maintaing version checks is a bit of a burden and looks like 1.10 will be the next LTS, what do you think about ditching LLVM 14 (Julia 1.9) support?

**Comment by mofeing** [comment]
The Brutus example works now with this PR on both Julia 1.10 and 1.11

File: src/IR/ExecutionEngine.jl [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+# TODO mlirExecutionEngineInvokePacked
+
+"""
+    lookup(jit, name)
+
+Lookup a native function in the execution engine by name, returns nullptr if the name can't be looked-up.
+"""
+function lookup(jit::ExecutionEngine, name::String; packed::Bool = false)
+    fn = packed ? API.mlirExecutionEngineLookupPacked(jit, name) : API.mlirExecutionEngineLookup(jit, name)
+    isnothing(fn) ? nothing : fn
```
Comment (Pangoraw) [inline src/IR/ExecutionEngine.jl:39]: did you mean `fn == C_NULL`?
```suggestion
    fn == C_NULL ? nothing : fn
```

**Pull request merged**